zebra, lib: new zebra TED module

mjstapp opened the pull request:
Draft for a new zebra module, currently called 'ted', to support some new TE/SR/LSDB functionality. The new module runs its own pthread, and is intended to be able to do a couple of things - eventually:
1) handle one or more zapi messages that are passed among daemons, but are opaque to zebra. The specific case is for topology and link information that will flow from IGPs to BGP (for BGP-LS), and possibly to other consumers of topology and path information.
2) possibly, depending on the ongoing design discussions, maintain a copy of the current state of the topology information itself.

The initial commits introduce the new zebra module, and start and stop its pthread in coordination with the zebra main pthread. The new module offers an api towards zserv that allows a batch of messages to enqueued towards it, much like the batch message management that's done for zclients' io pthreads currently. In order to allow zclient sessions to be accessed from the new ted pthread, there's some new busy- or ref-counting machinery that's intended to prevent the zebra main pthread from destroying a client session that's in-use.

qlyoung commented on zebra/zebra_ted.c:
```diff
@@ -0,0 +1,280 @@
+
+	frr_pthread_stop(zt_info.pthread, NULL);
+
+	THREAD_OFF(zt_info.t_msgs);
```
Not only is this unnecessary - you already killed the whole pthread and so the event loop isn't even running - it's also not thread safe even if it were running, you would need to use an async cancel.

qlyoung commented on zebra/zebra_ted.c:
```diff
@@ -0,0 +1,280 @@
+	bool need_resched = false;
+
+	/* Check for zebra shutdown */
+	if (!zt_info.run)
```
It would be better to check the `running` field in the `frr_pthread` struct itself, which has the same information. That check will need to be atomic as well.

qlyoung commented on zebra/zebra_ted.c:
```diff
@@ -0,0 +1,280 @@
+	/* Associate event 'master' */
+	zt_info.master = zt_info.pthread->master;
+
+	zt_info.run = true;
```
`frr_pthread` already has a `running` variable used for this purpose

qlyoung commented on zebra/zebra_ted.c:
```diff
@@ -0,0 +1,280 @@
+	atomic_fetch_add_explicit(&zt_info.msgs_in, i, memory_order_relaxed);
+
+	/* Check for zebra shutdown */
+	if (!zt_info.run) {
```
As before...

qlyoung commented on zebra/zebra_ted.h:
```diff
@@ -0,0 +1,58 @@
+ */
+bool zebra_ted_handle_msgid(uint16_t id);
+
+/* Module stop, called from the main pthread. This is synchronous:
```
Blank comment line above this one

qlyoung commented on zebra/zebra_ted.h:
```diff
@@ -0,0 +1,58 @@
+ */
+void zebra_ted_stop(void);
+
+/* Module cleanup, called from the zebra main pthread. When it returns,
```
Blank comment line above this one

qlyoung commented on zebra/zebra_ted.h:
```diff
@@ -0,0 +1,58 @@
+ */
+void zebra_ted_finish(void);
+
+/* Enqueue a batch of messages for processing. Returns the number dequeued
```
Blank comment line above this one

qlyoung commented on zebra/zserv.c:
```diff
@@ -629,26 +636,45 @@ static void zserv_client_free(struct zserv *client)
 
-	/* remove from client list */
-	listnode_delete(zrouter.client_list, client);
+	/* Final check in case the client struct is in use in another
```
Blank comment line above this one

qlyoung commented on zebra/zserv.c:
```diff
@@ -730,6 +758,60 @@ static struct zserv *zserv_client_create(int sock)
+			 * session is closed, schedule cleanup
+			 * on the zebra main pthread.
+			 */
+			if (client->is_closed)
```
Merge this check into the previous condition?

sworleys commented:
since we are adding new pthreads, probably a good time to start to document zebra's pthread architecture in the dev guide: http://docs.frrouting.org/projects/dev-guide/en/latest/zebra.html

mjstapp commented on lib/zclient.c:
```diff
@@ -2961,6 +2961,29 @@ static void zclient_mlag_handle_msg(ZAPI_CALLBACK_ARGS)
+	zclient_create_header(s, ZEBRA_OPAQUE_MESSAGE, VRF_DEFAULT);
+
+	/* Send opaque data */
+	stream_write(s, data, datasize);
```
oh, sorry - didn't realize until I looked just now at the way those asserts are set up. ouch.
can you say more about 'safe macro variants' - I only see a couple of `GET` macros. or are you suggesting adding a WRITE (or PUT) macro following the GET pattern? I'd be happy to do that.

mjstapp commented:
I think I've addressed the code comments up to this point now: added a `stream_fifo_deinit()` api, added opaque message-length checking, and some comment/style cleanups.

sworleys commented:
> I think I've addressed the code comments up to this point now: added a `stream_fifo_deinit()` api, added opaque message-length checking, and some comment/style cleanups.

Add to the doc guide? See my comment above.

mjstapp commented:
oh, yes, I have the doc open - just did a pass on the code itself this round, but I will get to the doc (and the 'zebra architecture' slide too).

> 
> Add to the doc guide? See my comment above.

odd22 commented on zebra/zebra_ted.c:
```diff
@@ -0,0 +1,270 @@
+/*
+ * Zebra TE data module
+ * Copyright (c) 2020 Volta Networks, Inc.
```
Align name of new module according to the name of new message. Perhaps `link-state` is more appropriate as information send by IGP are Link State info, not the Traffic Engineering Database

mjstapp commented:
Closing this in favor of #6236 - we agreed to rename the new zebra module (and apis).
